core/src/main/java/org/infinispan/persistence/spi/NonBlockingStore.java

```diff
@@ -80,114 +146,344 @@
+       * {@link #publishEntries(IntSet, Predicate, boolean)}. It is recommended that a store use the provided
+       * {@link org.infinispan.commons.time.TimeService} in the {@code InitializationContext} to determine if an
+       * entry has expired.
+       * <p>
+       * The information about an entry and its expiration is included in the {@link org.infinispan.metadata.Metadata}
+       * which is accessible from the {@link MarshallableEntry} which is provided.
        * <p>
        * Stores that have this characteristic must override the {@link #purgeExpired()} method.
        */
       EXPIRATION
    }
 
    /**
+    * The first method that will be invoked on this store to allow for the store to configure and do any additional
+    * steps, such as connecting via socket or opening file descriptors.
     * <p>
-    * This method is guaranteed to not be invoked concurrently with other operations. This also means another method
+    * The provided {@link InitializationContext} contains many helpful objects, including the configuration of the
+    * cache and store, concurrency utilities such as {@link org.infinispan.util.concurrent.BlockingManager} or
+    * an executor reserved for non blocking operations only {@link InitializationContext#getNonBlockingExecutor()}.
+    * <p>
+    * This method is guaranteed to not be invoked concurrently with other operations. This means another method
     * will not be invoked on this store until after the returned Stage completes.
+    * <p>
+    * It is expected that an implementation should be able to "restart" by invoking {@code start} a second time if
+    * {@link #stop()} has been invoked and allowed for its stage to complete.
     * @param ctx initialization context used to initialize this store
     * @return a stage that when complete signals that this store has been successfully started
     */
    CompletionStage<Void> start(InitializationContext ctx);
 
    /**
+    * This method will be invoked is being shut down. It is expected that all resources related to the store to be
+    * freed upon completion of the returned stage.
     * <p>
     * This method is guaranteed to not be invoked concurrently with other operations. This also means another method
     * will not be invoked on this store until after the returned Stage completes.
+    * <p>
+    * It is expected that an implementation should be able to "restart" by invoking {@link #start(InitializationContext)}
+    * a second time if {@code stop} has been invoked and allowed for its stage to complete.
     * @return a stage that when complete signals that this store has been stopped
     */
    CompletionStage<Void> stop();
 
    /**
-    *
-    * @return
+    * Returns a set of characteristics for this store and its elements. This method will be invoked one to many times
+    * to determine which methods of the store can be used and how its data can be handled.
```

Comment:
"This method will be invoked multiple times to ..."

Reply:
Tweaked to "This method may be invoked multiple times to ..."

Reply:
Although that doesn't convey it will be invoked at least one time, which is what I was going for before. WDYT?

Reply:
"One or more times"?

As long as the user knows to handle the fact that the method may be invoked multiple times, I don't think that it could be called just once really matters.